### PR TITLE
feat: support @computed decorator

### PR DESCRIPTION
**Pull request by tpoisseau**
Closes: https://github.com/zakodium/adonis-mongodb/issues/135

### CODE REVIEW

**Comment by targos** [comment]
Did you try to make an API like Lucid?

**Comment by tpoisseau** [comment]
Yes, I try.
I didn't found every places and it's hard to replicate code in two different codebase with differences.
The ActiveRecord pattern is very confusing by mixing db row instances with DB API, and metadata of collection.

But, I ended to be able to make a correct `@computed()` getter without breaking existent tests

**Comment by tpoisseau** [comment]
I added support on toJSON serialization, if you see something else Lucid support with `@computed` you found important, Say it to me please

File: src/Model/proxyHandler.ts [hunk 1]
```diff
@@ -1,11 +1,24 @@
...
+  get(target: any, prop: string | symbol, receiver: any) {
+    const Model = target.constructor as typeof BaseModel;
+    if (Model.$hasComputed(prop as string)) {
+      const property = Object.getOwnPropertyDescriptor(
+        Object.getPrototypeOf(target),
+        prop,
+      );
+      if (property?.get) {
+        return property.get.call(target.$attributes);
+      }
```
Comment (targos) [inline src/Model/proxyHandler.ts:14]: In Lucid, it's supposedly different. I think the getter ends up being called on the Proxy, so you are able to access attributes, but also methods on the model.

Comment (tpoisseau) [reply src/Model/proxyHandler.ts:14]: I tried different combinaisons (with receiver argument of proxy handler), but only this one works without breaking other tests

Comment (tpoisseau) [reply src/Model/proxyHandler.ts]: Ok, I retried with Reflect.get and it's working now.

**Comment by tpoisseau** [comment]
CleanShot 2023-11-09 at 08 54 55

File: src/Model/Model.ts [hunk 2]
```diff
@@ -622,7 +665,18 @@ export class BaseModel {
...
 
   public toJSON(): unknown {
-    return this.$attributes;
+    const Model = this.constructor as typeof BaseModel;
+
+    const computed: Record<string, unknown> = {};
+    for (const key of Model.$computedDefinitions.keys()) {
+      // @ts-expect-error polymorphic getter
+      computed[key] = this[key];
+    }
```
Comment (targos) [inline src/Model/Model.ts:674]: You need to ignore keys if `serializeAs` is `null`

**Review by targos — APPROVED** [review]
LGTM after fix to `serializeAs: null`

File: src/Model/Model.ts (hunk 2 above)
Comment (tpoisseau) [reply src/Model/Model.ts]: Done, full support of serializeAs.

* `null` ignore serialization
* non-empty string use it as key
* falsy value fallback on getter name

**Pull request merged**